### ISSUE TITLE
Symbol takes half a second to demangle with default configuration

### ISSUE THREAD

**Issue by athre0z** [gimli-rs/cpp_demangle#282]
We encountered the following symbol in a real-world executable:

```
_ZN13QConcatenableI14QStringBuilderIS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_IS0_I7QStringS1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_ES1_EE8appendToI5QCharEEvRKS15_RPT_
```

GNU `c++filt` swiftly decodes it to the following in less than one millisecond:

```
void QConcatenable<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QString, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString> >::appendTo<QChar>(QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QStringBuilder<QString, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString>, QString> const&, QChar*&)
```

An LTOed release build using `cpp_demangle` with default options running on an M1 Max munches on it for almost half a second before giving up with 

```
cpp_demangle::BorrowedSymbol::new(name.as_bytes()) = Err(
    UnexpectedText,
)
```

Code:

```rust
if let Ok(sym) = cpp_demangle::BorrowedSymbol::new(name.as_bytes()) {
    let mut demangled = String::new();
    let options = cpp_demangle::DemangleOptions::default();
    if let Ok(()) = sym.structured_demangle(&mut demangled, &options) {
        dbg!(demangled);
    }
}
```

**Comment by khuey** [comment]
This pathological symbol triggers the default recursion limits in cpp_demangle. The error handling paths are slow (I haven't investigated that) and lead to the 0.5 second runtime you see. If you bump the recursion limits (I haven't attempted to find minimal values here) like so, it will parse correctly in less than a tenth of that time.

```rust
if let Ok(sym) = cpp_demangle::BorrowedSymbol::new_with_options(name.as_bytes(),  &ParseOptions::default().recursion_limit(256)) {
    let mut demangled = String::new();
    let options = cpp_demangle::DemangleOptions::default()
        .recursion_limit(256);
    if let Ok(()) = sym.structured_demangle(&mut demangled, &options) {
        dbg!(demangled);
    }
}
```

**Comment by athre0z** [comment]
Thanks a lot for the swift fix -- works great here! The pathological executable where it used to take more than an hour to demangle all the symbols now parses in 3 seconds.

**Comment by khuey** [comment]
Was that with the recursion limit at the default?

**Comment by athre0z** [comment]
Yes.

**Comment by khuey** [comment]
Great!

**Comment by khuey** [comment]
I released 0.4.2 with this fixed.
